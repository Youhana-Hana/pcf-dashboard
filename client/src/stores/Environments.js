import AppDispatcher from '../dispatcher/AppDispatcher';
import Constants from '../constants/PcfPipelines.js';
import { EventEmitter } from 'events';

const CHANGE_EVENT  = 'change';

let _environments = [];

function setEnvironemnts(environments) {
    _environments = environments;
}

class EnvStoreClass extends EventEmitter {

    emitChange() {
        this.emit(CHANGE_EVENT);
    }

    addChangeListener(callback) {
        this.on(CHANGE_EVENT, callback);
    }

    removeChangeListener(callback) {
        this.removeListener(CHANGE_EVENT, callback);
    }

    getEnvironments() {
        return _environments;
    }
}

const EnvStore = new EnvStoreClass();

EnvStore.dispatchToken = AppDispatcher.register(action => {
    switch(action.actionType) {
    case Constants.RECIEVE_PAYLOAD:
        setEnvironemnts(action.data.environments);
        EnvStore.emitChange();
    }
});

export default EnvStore;
