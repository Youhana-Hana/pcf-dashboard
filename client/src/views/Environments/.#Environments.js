developer@ubuntu-dev-001.7455:1509615081