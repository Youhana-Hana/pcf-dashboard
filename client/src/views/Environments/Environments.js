import React, { Component } from 'react';
import {
  Row,
  Col,
  Card,
  CardHeader,
  CardBlock,
} from "reactstrap";

import Environment from '../../components/Environment/';
import EnvStore from '../../stores/Environments';
import PcfAutomationStatusActions from '../../actions/PcfAutomationStatus.js';
import PcfAutomationStatusConstants from '../../constants/PcfAutomationStatus.js';

const API_URL = '/metrics/environments';

class Environments extends Component {
  constructor(props) {
    super(props);
    this.state =  {
      environments: []
    };

    this.onChange = this.onChange.bind(this);
  }

  tick() {
    PcfAutomationStatusActions.loadPcfAutomationStatus(API_URL, PcfAutomationStatusConstants.RECEIVE_PAYLOAD_ENVIRONMENTS);
  }

  componentDidMount() {
    PcfAutomationStatusActions.loadPcfAutomationStatus(API_URL, PcfAutomationStatusConstants.RECEIVE_PAYLOAD_ENVIRONMENTS);

  }

  componentWillMount() {
    EnvStore.addChangeListener(this.onChange);
    this.interval = setInterval(this.tick, 10000);
  }

  componentWillUnmount() {
    EnvStore.removeChangeListener(this.onChange);
    clearInterval(this.interval);
  }


  onChange() {
    this.setState({
      environments: EnvStore.getEnvironments()
    });
  }
  render() {
    return (
      <Row>
        <Col>
          <Card>
            <CardHeader>
              <h1 className="mb-0">Environments</h1>
            </CardHeader>
            <CardBlock className="card-body">
              <Row>

                {
                  this.state.environments.map((env) => {
                    return <Environment env={env} key={env.region+"-"+env.foundation}></Environment>;
                  })
                }
              </Row>
            </CardBlock>
          </Card>
        </Col>
      </Row>
    )}
}

export default Environments;
