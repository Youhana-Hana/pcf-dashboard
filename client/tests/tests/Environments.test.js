import { shallow, mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import Environments  from '../../src/views/Environments/Environments';
import EnvironmentStore from '../../src/stores/Environments'
import toJson from 'enzyme-to-json';

jest.mock('../../src/stores/Environments.js');

describe('Environments', () => {
  const environments = [{
    "id": "us-t3",
    "foundation": "T3",
    "region": "US",
    "products": {
      "name": "elastic-runtime",
      "currentVersion": "1.10.13",
      "stagedVersion": "",
      "currentVersionInS3": "1.10.18",
      "pcfPipelineVersion": "1.15.1",
      "pipelineStatus": "failed",
      "buildInfo": {
        "pipelineName": "upgrade-tile-ert",
        "buildNumber": "22",
        "buildGlobalIdentifier": "56",
        "buildUrl": "http://myconcourse-on-us-t3.com/builds/56",
        "failedJobName": "configure-ert"
      }
    }
  }, {
      "id": "us-prod",
      "foundation": "prod",
      "region": "US",
      "products": {
        "name": "elastic-runtime",
        "currentVersion": "1.10.13",
        "stagedVersion": "",
        "currentVersionInS3": "1.10.18",
        "pcfPipelineVersion": "1.15.1",
        "pipelineStatus": "failed",
        "buildInfo": {
          "pipelineName": "upgrade-tile-ert",
          "buildNumber": "22",
          "buildGlobalIdentifier": "56",
          "buildUrl": "http://myconcourse-on-us-prod.com/builds/56",
          "failedJobName": "configure-ert"
        }
      }
    }]

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Environments/>,
    );

    const callback = jest.fn();

    EnvironmentStore.getEnvironments = jest.fn(() => {
      return environments;
    })

    EnvironmentStore.addChangeListener = jest.fn((callback) => {
      callback();
    })

    EnvironmentStore.removeChangeListener = jest.fn((callback) => {
      callback();
    })
  });

  afterEach(() => {
    EnvironmentStore.getEnvironments.mockClear();
    EnvironmentStore.addChangeListener.mockClear();
    EnvironmentStore.removeChangeListener.mockClear();
  });

  it('should render environemnts header', () => {
    const label= wrapper.find('h1').html();

    expect(label).toEqual('<h1 class="mb-0">Environments</h1>');
  })

  it('should render first environment', () => {
    wrapper.setState({environments: environments});

    const t4= wrapper.find('Environment').first();

    expect(t4.props().env).toEqual(environments[0]);
    expect(t4.key()).toEqual(environments[0].id);
  })

  it('should render last environment', () => {
    wrapper.setState({environments: environments});

    const t5= wrapper.find('Environment').last();

    expect(t5.props().env).toEqual(environments[1]);
    expect(t5.key()).toEqual(environments[1].id);
  })

  it('test to see if the environments renders correctly', () => {
    wrapper.setState({environments: environments});

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('register to EnvironmentStore', () => {
    wrapper = mount(
      <Environments/>,
    );

    expect(EnvironmentStore.addChangeListener.mock.calls.length).toEqual(1);
  })

  it('unRegister to EnvironmentStore', () => {
    wrapper = mount(
      <Environments/>,
    );

    wrapper.unmount();
    expect(EnvironmentStore.removeChangeListener.mock.calls.length).toEqual(1);
  })
});
