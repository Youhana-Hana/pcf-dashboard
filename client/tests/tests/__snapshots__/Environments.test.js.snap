// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environments test to see if the environments renders correctly 1`] = `
<Row
  tag="div"
>
  <Col
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <Card
      tag="div"
    >
      <CardHeader
        tag="div"
      >
        <h1
          className="mb-0"
        >
          Environments
          <div
            className="align-bottom float-right"
          >
            <Spinner
              className="align-middle"
              loading={false}
              size="15"
            />
          </div>
        </h1>
      </CardHeader>
      <CardBlock
        className="card-body"
        tag="div"
      >
        <Row
          tag="div"
        >
          <Environment
            env={
              Object {
                "foundation": "T3",
                "id": "US-T3",
                "products": Object {
                  "buildInfo": Object {
                    "buildGlobalIdentifier": "56",
                    "buildNumber": "22",
                    "buildUrl": "http://myconcourse-on-us-t3.com/builds/56",
                    "failedJobName": "configure-ert",
                    "pipelineName": "upgrade-tile-ert",
                  },
                  "currentVersionERT": "1.10.13",
                  "currentVersionERTInS3": "1.10.18",
                  "name": "elastic-runtime",
                  "pcfPipelineVersion": "1.15.1",
                  "pipelineStatus": "failed",
                  "stagedVersionERT": "",
                },
                "region": "US",
              }
            }
            key="US-T3"
          />
          <Environment
            env={
              Object {
                "foundation": "prod",
                "id": "US-prod",
                "products": Object {
                  "buildInfo": Object {
                    "buildGlobalIdentifier": "56",
                    "buildNumber": "22",
                    "buildUrl": "http://myconcourse-on-us-prod.com/builds/56",
                    "failedJobName": "configure-ert",
                    "pipelineName": "upgrade-tile-ert",
                  },
                  "currentVersionERT": "1.10.13",
                  "currentVersionERTInS3": "1.10.18",
                  "name": "elastic-runtime",
                  "pcfPipelineVersion": "1.15.1",
                  "pipelineStatus": "failed",
                  "stagedVersionERT": "",
                },
                "region": "US",
              }
            }
            key="US-prod"
          />
        </Row>
      </CardBlock>
    </Card>
  </Col>
</Row>
`;
